### PR TITLE
ARTEMIS-1422 fix RedeployTest after match change

### PR DESCRIPTION
**Pull request by jbertram**
Commit 8bd8e2bd5d4fc2fc3e69bcfbb20799bc55bc1e9e for ARTEMIS-1403
broke RedeployTest because it wasn't using the proper delimiter
in the address/queue names and matches.

### CODE REVIEW

**Comment by michaelandrepearce** [comment]
This shouldn't be affected, if that change affects this than there's issue in match change

**Comment by jbertram** [comment]
The test org.apache.activemq.artemis.tests.integration.jms.RedeployTest#testRedeployAddressQueue succeeds before 8bd8e2bd5d4fc2fc3e69bcfbb20799bc55bc1e9e and then fails afterward so regardless of whether or not it *should* be affected, it is affected.

At first I also thought the change in 8bd8e2bd5d4fc2fc3e69bcfbb20799bc55bc1e9e was wrong, but looking at it more closely I believe RedeployTest was actually in error for not using the proper delimiter for matching.  [The documentation](http://activemq.apache.org/artemis/docs/latest/wildcard-syntax.html) states:

> ...wildcard expression contains words delimited by the character '.' (full stop)...The special characters '#' and '*' also have special meaning and can take the place of a word.

In other words, the special characters '#' and '*' match words and words are delimited with '.' characters.

From what I can tell all the examples in the documentation and the code use the '.' delimiter appropriately except for RedeployTest which uses '\_' instead.  Before the fix in 8bd8e2bd5d4fc2fc3e69bcfbb20799bc55bc1e9e using '\_' worked since '#' was transformed into '.*' which would match any string without respect to the delimiter.  Once the code was fixed to respect the delimiter then the test began to fail.

**Comment by michaelandrepearce** [comment]
So if anything only this should be needed to be added. 

+      <wildcard-addresses>
 +        <enabled>true</enabled>
 +        <delimiter>_</delimiter>
 +        <any-words>#</any-words>
 +        <single-word>*</single-word>
 +      </wildcard-addresses>

**Comment by jbertram** [comment]
Configuring the delimiter to be '\_' was my second thought.  Unfortunately that doesn't appear to be implemented at this point as the test still failed.  That configuration appears to work properly for *routing* but not for adjusting the matching logic.

**Comment by jbertram** [comment]
BTW, the test **is** reliant on wild-card matching as it's required to apply these address-settings:

            <config-delete-queues>FORCE</config-delete-queues>
            <config-delete-addresses>FORCE</config-delete-addresses>

If those address settings are not applied then the test fails and since the "match" for the address setting was originally "config\_#" the test fails (since '\_' is not a valid delimiter).

**Comment by michaelandrepearce** [comment]
So if this test breaks now, then theres something wrong with the matching change causing problems with non wildcards, or where other wildcard separators can be used.

This will affect users, who do not use '.' separator due to over issues with its over use and expansion else where (inc the org i work for)

**Comment by michaelandrepearce** [comment]
I would suggest that change is reverted, and this is discussed more with that change, or further work is done in that change to properly handle wildcard config.

**Comment by jbertram** [comment]
I agree that there's still a problem, but I don't think the matching change should be reverted.  The matching logic was *not* working as advertised before the change.  Your test-case was simply relying on the broken behavior.

As I mentioned previously I would suggest that the delimiter (and other characters) be configurable for the matching logic as well as the routing.  This was probably the intended functionality all along when ARTEMIS-815 was implemented.

**Comment by michaelandrepearce** [comment]
I fully disagree.

**Comment by jbertram** [comment]
The matching logic change isn't breaking anybody at this point unless they are using daily snapshots of the master branch or something (which isn't recommended in the first place).  It's not in the latest release (i.e. 2.3.0).  As long as the matching logic is changed to support (at the very least) a configurable delimiter before the next release then nobody will ultimately be broken.  IMO reverting 8bd8e2bd5d4fc2fc3e69bcfbb20799bc55bc1e9e now and then re-committing it later solves nothing.

**Comment by michaelandrepearce** [comment]
Thats fair comment, as long as we do, though to ensure that then should leave this test failing, and prioritise making it work with just wildcard config being applied.

**Comment by jbertram** [comment]
Or we could just create a JIRA and set it as a blocker on the next release.

**Comment by michaelandrepearce** [comment]
Id rather we fix asap. As said I'm happy to try do this.

**Comment by clebertsuconic** [comment]
@michaelandrepearce You ok to merge this?

**Comment by michaelandrepearce** [comment]
@clebertsuconic have worked on a fix for the mentioned match change so we don't break users using custom wildcard config, essentially making the matcher correctly use the wildcard config.

https://github.com/apache/activemq-artemis/pull/1537

**Comment by clebertsuconic** [comment]
@jbertram I closed this one since @michaelandrepearce sent another fix.

let me know if you have any considerations about it please.
